src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -57,6 +56,106 @@
 #include "utilities/events.hpp"
 #include "utilities/preserveException.hpp"
 
+void MonitorList::add(ObjectMonitor* m) {
+  for (;;) {
+    ObjectMonitor* head = Atomic::load(&_head);
+    m->set_next_om(head);
+    if (Atomic::cmpxchg(&_head, head, m) == head) {
+      break;
+    }
+  }
+  size_t count = Atomic::add(&_count, 1u);
+  if (count > max()) {
+    Atomic::inc(&_max);
+  }
+}
+
+size_t MonitorList::count() const {
+  return Atomic::load(&_count);
+}
+
+size_t MonitorList::max() const {
+  return Atomic::load(&_max);
+}
+
+// Walk the in-use list and unlink (at most MonitorDeflationMax) deflated
+// ObjectMonitors. Returns the number of unlinked ObjectMonitors.
+size_t MonitorList::unlink_deflated(Thread* self, LogStream* ls,
+                                    elapsedTimer* timer_p,
+                                    GrowableArray<ObjectMonitor*>* unlinked_list) {
+  size_t unlinked_count = 0;
+  ObjectMonitor* prev = NULL;
+  ObjectMonitor* head = Atomic::load_acquire(&_head);
+  ObjectMonitor* m = head;
+  while (m != NULL) {
+    if (m->is_being_async_deflated()) {
+      // Find next live ObjectMonitor.
+      ObjectMonitor* next = m;
+      while (next != NULL && next->is_being_async_deflated()) {
+        ObjectMonitor* next_next = next->next_om();
+        unlinked_count++;
+        unlinked_list->append(next);
+        next = next_next;
+        if (unlinked_count >= (size_t)MonitorDeflationMax) {
+          // Reached the max so bail out on the gathering loop.
+          break;
+        }
+      }
+      if (prev == NULL) {
+        ObjectMonitor* prev_head = Atomic::cmpxchg(&_head, head, next);
+        if (prev_head != head) {
+          // Find new prev ObjectMonitor that just got inserted.
+          for (ObjectMonitor* n = prev_head; n != m; n = n->next_om()) {
+            prev = n;
+          }
+          prev->set_next_om(next);
+        }
+      } else {
+        prev->set_next_om(next);
+      }
+      if (unlinked_count >= (size_t)MonitorDeflationMax) {
+        // Reached the max so bail out on the searching loop.
+        break;
+      }
+      m = next;
+    } else {
+      prev = m;
+      m = m->next_om();
+    }
+
+    if (self->is_Java_thread() &&
```

Comment:
Non-JavaThreads may have a monitor list ???

Reply:
This function, MonitorList::unlink_deflated(), may be called by
either the MonitorDeflationThread or the VMThread. The VMThread
does not need to block for the safepoint which is what the
`if (self->is_Java_thread()` prevents.